k2/csrc/array.h

```diff
@@ -139,9 +139,13 @@ struct Array2 {
   */
   bool Empty() const { return size1 == 0; }
 
-  PtrT begin() const { return data + indexes[0]; }
+  PtrT begin() { return size2 != 0 ? data + indexes[0] : nullptr; }
```

Comment:
It might be worthwhile to let `indexes` point to a zero (e.g. could use the address of `size` itself) in the case where size == 0.  Would avoid if-statements, which can be slow.

Reply:
Updated, thanks!